TestCentricEngine.md

```diff
@@ -0,0 +1,82 @@
+
+ * **Agent:** Test __agents__ form an important part of the __engine layer__. Agents are responsible for actually discovering and executing tests. They are logically distinct, however, because they generally need to be built for a greater variety of platforms than does the __engine assembly__.
+
+ * **Framework:** A **test framework** is designed to run tests written agains a particular API. The **framework** is generally the only component that is referenced by the users test assembly. Frameworks represent the lowest of the three architectural layers.
+
+## NUnit Test Engine
+
+The current NUnit Test Engine layer is an implementation of the original design, with some changes made as time has gone on. It consists of the following assemblies:
+ * **nunit.engine.api.dll** defines the interfaces used by runners to access the engine. It is maintained at the same 3.0.0.0 version even as other assemblies are re-versioned and is only modified by adding new things.
+ * **nunit.engine.dll** is the implementation of the engine. It also contains the implementation of cross-process and in-process agents and their associated runners.
+ * **nunit-agent.exe** and **nunit-agent-x86.exe** are console apps that serve as cross-processs agents when invoked by the engine. They reference the engine code and use configuration options to enable only those parts of the engine assembly that are intended for use by an agent.
+
+It may seem odd that the NUnit engine assembly is used for two purposes: to function as an engine and to provide services as part of an agent process. This was initially done for expediency, with the plan of splitting the code into two assemblies, one related to engine functions _per se_ and the other containing the code needed by an agent. Unfortunately, this is a rather large amount of technical debt that was never paid off. 
+
+## TestCentric GUI Requirements
+
+Currently, the TestCentric GUI is using a modified build of the NUnit engine. So far, the modifications are minor. In this section, I want to examine some key features we will need in the near future, which are not currently provided by the NUnit engine.
+
+1. The Gui should be able to run tests on other platforms than .NET Framework, such as .NET Core. It should, in fact, be able to load and run multiple assemblies that use different platforms. Currently, that's not possible because it requires using two different builds of the engine in the same process.
```

Comment:
I think this is made very difficult by the insistence that we need .NET Standard builds of the engine. Rob and I had discussed an alternate architecture some years back. It's the architecture I am following in my spike of a new engine. 

The "engine" per se only has to run in limited platforms. Those on which developers do their work. It's only the "bottom half" of the current engine that needs to run against all the platforms for which tests are written.

My understanding is that this was a non-reversible decision, which is why I called this one a blocker. 

FWIW, the splitting of the "engine" into upper and lower (engine and agent) layers was part of the original plans for its development.

All that said, if the NUnit engine under one platform could develop to the point of running tests under all platforms, that would change the picture for me.

Reply:
For running tests for multiple platforms from the GUI (or the console for that matter), the current plan is to use the full framework engine and launch an appropriate NUnit Agent the same way we do it for x86. We'll need to tackle the IPC channel because .NET Core doesn't support Remoting.

Creating a .NET Standard version of 'half' of the engine was required to support .NET Core in the Adapter and with the `dotnet test` command. You wouldn't and can't use this build in a full framework runner, the intention is to use the full framework engine. We just haven't completed the work yet.

In your fork of the engine, you removed the process runner. How are you planning on loading .NET Core based test assemblies without a separate runner process? You cannot load and run .NET Core assemblies in a Full Framework process. Some things may appear to work, but you will quickly run into `MissingMethodException`s because of differences in the APIs.

Reply:
@rprouse Thanks for commenting! 

There are a few things I don't understand, however.

"Creating a .NET Standard version of 'half' of the engine was required..."

That's not what you have done. The entire engine is currently built for all platforms. Of course, through conditional compilation, there is a great deal of functionality that is not included in the .NET Standard builds, but it's still the the engine. In my view, this is a problem.

"In your fork of the engine, you removed the process runner."

Not clear what you are looking at. Although I have a spike, which we were discussing in this issue, I haven't published it anywhere simply __because__ it's a spike. It definitely has a ProcessRunner, for the obvious reason you state. I do know that .NET Core can't run under the .NET Framework.

All the public builds of the engine that you may have seen are intended for use with version 1.0 of TestCentric only and share the same limitations as the NUnit 3 Console Runner. They make no use of the .NET Standard builds, although I'm forced to build them if I want to stay compatible.

In the private (spike) build, ProcessRunner lives in the engine while most other runners live in what I'm tentatively calling nunit.agents.dll. The former needs only to be built for the platforms developers work under. I'm currently thinking I'll eventually support some version of .NET Framework 4.5 or higher and .NET Core 3.0. Of course, if this were done for the NUnit project rather than only for TestCentric, .NET Framework Framework 2.0 would still be needed unless you are willing to give up in-process execution.

The agents assembly is different. It has to be built for all the target frameworks we want to support for test execution. This difference in platform targets is the main reason I advocated for a split way back - unfortunately leading to my no longer working on the NUnit engine.

Reply:
@rprouse hasn't gotten back on the response to his comment, but I think we can move on. If the goal of both projects is to run all test platforms from a single engine, that's what's important.

So, the statement that it's not __currently__ possible is true, but we acknowledge that that can change.